### PR TITLE
Alternative approach to a QGIS specific unittest class

### PR DESCRIPTION
**Pull request by nyalldawson**
Possible alternative to https://github.com/qgis/QGIS/pull/53374

### CODE REVIEW

**Comment by nyalldawson** [comment]
@m-kuhn what do you think of this approach? It removes the patching entirely, which is a bit cleaner, but I'm not sure if there's unwanted side effects from this...

**Review by m-kuhn — APPROVED** [review]
It makes things more explicit and clear, I like it a lot.
Thanks :heart:

**Comment by nyalldawson** [comment]
@m-kuhn 

Thanks! So I guess one potential issue is if downstream projects relied on the additions to this class, like assertLayerEqual/assertGeometriesEqual. Do you have any thoughts on how we could/should handle that?

potentially I could leave the old additions to unittest.TestCase, and then make QgisTestCase(unitTest.TestCase) which adds the new report / image check handling only.

**Comment by m-kuhn** [comment]
Maybe this?

````
from warnings import warn

def assertGeometriesEqualDeprecated(self, a, b):
    warn('unittest.assertGeometriesEqual() is deprecated and will be removed in the future. Port your tests to `qgis.testing.TestCase`')
    return QgisTestCase.assertGeometriesEqual(self, a, b)

unittest.TestCase.assertGeometriesEqual = assertGeometriesEqualDeprecated
```

**Comment by m-kuhn** [comment]
@nyalldawson I hope you don't mind that I was so free as to push the proposal directly.

**Comment by m-kuhn** [comment]
Ehm...


```
<QgsCoordinateReferenceSystem: EPSG:4326> !=<QgsCoordinateReferenceSystem: EPSG:4326>
```

**Comment by m-kuhn** [comment]
@nyalldawson I am blocked with this, do you have any idea what could be going wrong?

File: python/testing/__init__.py [hunk 1]
```diff
@@ -221,13 +220,13 @@ def checkLayersEqual(self, layer_expected, layer_result, use_asserts=False, **kw
             result_wkt = layer_result.dataProvider().crs().toWkt(QgsCoordinateReferenceSystem.WKT_PREFERRED)
 
             if use_asserts:
-                _TestCase.assertEqual(self, layer_expected.dataProvider().crs(), layer_result.dataProvider().crs())
+                self.assertEqual(self, layer_expected.dataProvider().crs(), layer_result.dataProvider().crs())
```
Comment (nyalldawson) [inline python/testing/__init__.py:223]:
```suggestion
                self.assertEqual(layer_expected.dataProvider().crs(), layer_result.dataProvider().crs())
```

File: python/testing/__init__.py [hunk 2]
```diff
@@ -221,13 +220,13 @@ def checkLayersEqual(self, layer_expected, layer_result, use_asserts=False, **kw
...
             elif layer_expected.dataProvider().crs() != layer_result.dataProvider().crs():
                 return False
 
         # Compare features
         if use_asserts:
-            _TestCase.assertEqual(self, layer_expected.featureCount(), layer_result.featureCount())
+            self.assertEqual(self, layer_expected.featureCount(), layer_result.featureCount())
```
Comment (nyalldawson) [inline python/testing/__init__.py:229]:
```suggestion
            self.assertEqual(layer_expected.featureCount(), layer_result.featureCount())
```

**Comment by nyalldawson** [comment]
Hopefully [1fc2e2b](https://github.com/qgis/QGIS/pull/53403/commits/1fc2e2bc426e2115e1627f151136c27093576415) fixes it

**Comment by m-kuhn** [comment]
Thank you for pushing this through the finish line :heart:

**Pull request merged**